File: src/math/rectangle.ts
import { compare, max, min } from "./number";
import Vector2 from "./vector2";

class Rectangle {
  public readonly x: number;
  public readonly y: number;
  public readonly width: number;
  public readonly height: number;

  public constructor(x: number, y: number, width: number, height: number) {
    this.x = x;
    this.y = y;
    this.width = width;
    this.height = height;
  }

  public getIntersection(other: Rectangle): Rectangle | undefined {
    const x = max(this.x, other.x);
    const y = max(this.y, other.y);
    const x1 = min(this.x + this.width, other.x + other.width);
    const y1 = min(this.y + this.height, other.y + other.height);

    const width = x1 - x;
    const height = y1 - y;

    if (compare(width, 0) < 0 || compare(height, 0) < 0) return undefined;

    return new Rectangle(x, y, width, height);
  }

  public containsPoint(point: Vector2): boolean {
    return (
      compare(point.x, this.x) >= 0 &&
      compare(point.y, this.y) >= 0 &&
      compare(point.x, this.x + this.width) <= 0 &&
      compare(point.y, this.y + this.height) <= 0
    );
  }
}

export default Rectangle;

File: logic/src/math/rectangle.rs
use num_traits::Num;

use super::{compare, Polygon, Vector2};

#[derive(Debug, Default, Clone)]
pub struct Rectangle<T: Num + Copy> {
    pub x: T,
    pub y: T,
    pub width: T,
    pub height: T,
}

impl<T: Num + Copy> Rectangle<T> {
    pub const fn new(x: T, y: T, width: T, height: T) -> Self {
        Self {
            x,
            y,
            width,
            height,
        }
    }

    pub fn from_center(center: &Vector2<T>, width: T, height: T) -> Self {
        let two = T::one() + T::one();

        Self {
            x: center.x - width / two,
            y: center.y - height / two,
            width,
            height,
        }
    }
}

impl Rectangle<f32> {
    pub fn contains_point(&self, point: &Vector2<f32>) -> bool {
        compare(point.x, self.x).is_ge()
            && compare(point.y, self.y).is_ge()
            && compare(point.x, self.x + self.width).is_le()
            && compare(point.y, self.y + self.height).is_le()
    }
}

impl From<&Rectangle<f32>> for Polygon {
    fn from(value: &Rectangle<f32>) -> Self {
        Polygon::new([
            Vector2::new(value.x, value.y),
            Vector2::new(value.x, value.y + value.height),
            Vector2::new(value.x + value.width, value.y + value.height),
            Vector2::new(value.x + value.width, value.y),
        ])
    }
}
